src/Roslyn.Diagnostics.Analyzers/Core/AbstractApplyTraitToClass.cs

```diff
@@ -0,0 +1,111 @@
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Roslyn.Diagnostics.Analyzers
+{
+    public abstract class AbstractApplyTraitToClass<TAttributeSyntax> : CodeRefactoringProvider
+        where TAttributeSyntax : SyntaxNode
+    {
+        protected AbstractApplyTraitToClass()
+        {
+        }
+
+        private protected abstract IRefactoringHelpers RefactoringHelpers { get; }
+
+        protected abstract SyntaxNode? GetTypeDeclarationForNode(SyntaxNode reportedNode);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var attribute = await context.TryGetRelevantNodeAsync<TAttributeSyntax>(RefactoringHelpers).ConfigureAwait(false);
+            if (attribute is null)
+            {
+                // No attribute in context
+                return;
+            }
+
+            var syntaxGenerator = SyntaxGenerator.GetGenerator(context.Document);
+            if (syntaxGenerator.TryGetContainingDeclaration(attribute, DeclarationKind.Method) is null)
+            {
+                // The attribute is not applied to a method
+                return;
+            }
+
+            var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+            var attributeType = semanticModel!.GetTypeInfo(attribute, context.CancellationToken);
+            if (attributeType.Type is not { Name: "TraitAttribute", ContainingNamespace: { Name: "Xunit", ContainingNamespace: { IsGlobalNamespace: true } } })
+                return;
+
+            var location = attribute.GetLocation();
+            context.RegisterRefactoring(
+                CodeAction.Create(
+                    RoslynDiagnosticsAnalyzersResources.ApplyTraitToContainingType,
+                    cancellationToken => ApplyTraitToClassAsync(context.Document, location.SourceSpan, cancellationToken),
+                    nameof(AbstractApplyTraitToClass<TAttributeSyntax>)));
+        }
+
+        private async Task<Document> ApplyTraitToClassAsync(Document document, TextSpan sourceSpan, CancellationToken cancellationToken)
+        {
+            var semanticModel = (await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false))!;
+
+            var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxRoot = await syntaxTree!.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var reportedNode = syntaxRoot.FindNode(sourceSpan, getInnermostNodeForTie: true);
+            var typeDeclaration = GetTypeDeclarationForNode(reportedNode);
+            if (typeDeclaration is null)
+                return document;
+
+            var syntaxGenerator = SyntaxGenerator.GetGenerator(document);
+            if (syntaxGenerator.TryGetContainingDeclaration(reportedNode, DeclarationKind.Attribute) is not { } attribute)
```

Comment:
This code seems to differ from Register callback where we use TryGetRelevantNodeAsync. Is it possible to refactor and share code which tries to go from context to containing method?

Reply:
`TryGetRelevantNodeAsync` resolves a position or span to a syntax node with a specific type. When the refactoring implementation is called, we are no longer operating on the caret position but instead are operating on the span already resolved by `TryGetRelevantNodeAsync`.